Replace fpu reduction with best child eval

WhiteHalmos opened the pull request:
* Use minimum of best visited child's eval and parent eval when a node
has not been visited.
* Update AUTHORS (to be compliant with my company policy which allows me to contribute).
* Based on @Eddh's totalvp work in #827.

WhiteHalmos commented:
Current results against next (fpu - 0.22):
```
LZ-next v LZ-next-visited-best-eval (61/400 games)      
board size: 19   komi: 7.5  
                            wins              black         white       avg cpu                                  
LZ-next                       31 50.82%       15 48.39%     16 53.33%    256.14                                  
LZ-next-visited-best-eval     30 49.18%       14 46.67%     16 51.61%    273.22                                  
                                              29 47.54%     32 52.46%                                            
```

Starting to test with oldest network 1ac2 against next (fpu - 0).

<details>
<summary>Gomill config</summary>

```
competition_type = 'playoff'
LZ_ARGS = '/projects/lztest/{} -p1600 --noponder -r1 -g -t1 -q -d -w {} {}'
def LZ(binary, weights, extra_args=''):
  return Player(LZ_ARGS.format(binary, weights, extra_args))
players = dict()
p1 = 'LZ-next'
p2 = 'LZ-next-visited-best-eval'
players[p1] = LZ('leelaz-next-0dd0', 'd6')
players[p2] = LZ('leelaz-0cb8', 'd6')
board_size = 19
komi = 7.5
matchups = [Matchup(p1, p2, alternating=True, number_of_games=400)]
```
</details>

Eddh commented:
Interesting. That was one of the idea I had at the very start. It didn't seem to work as well as -0.2 but that could have been the very bad sample size I used. Or I implemented it badly and it was bugged somehow ? It would have been an achievement to fuck up something that simple even for me though. If it actually works that's the dream I had for fpu come true. Thanks !

WhiteHalmos commented:
I'm hopeful that this holds water, but only testing and more data will tell. Here's what I'm running right now against next (fpu - 0) on 1ac2:

```
LZ-next-fpu0 v LZ-next-visited-best-eval (1/400 games)  
board size: 19   komi: 7.5  
                            wins                   avg cpu                                                       
LZ-next-fpu0                   1 100.00%   (black)  102.59                                                       
LZ-next-visited-best-eval      0   0.00%   (white)  101.68
```

<details>
<summary>Gomill config2</summary>

```
competition_type = 'playoff'
LZ_ARGS = '/projects/lztest/{} -p1600 --noponder -r1 -g -t1 -q -d -w {} {}'
def LZ(binary, weights, extra_args=''):
  return Player(LZ_ARGS.format(binary, weights, extra_args))
players = dict()
p1 = 'LZ-next-fpu0'
p2 = 'LZ-next-visited-best-eval'
players[p1] = LZ('leelaz-next-0dd0', '1a', '--fpu_reduction 0.0')
players[p2] = LZ('leelaz-nofpu-2ec8', '1a')
board_size = 19
komi = 7.5
matchups = [Matchup(p1, p2, alternating=True, number_of_games=400)]
```
</details>

gcp commented:
I like the elegance of this approach (no magic number).

WhiteHalmos commented:
Rebased and updated games. Seems it is about even on the latest net (6665) with next (fpu - 0.22) and slightly worse on one of the oldest nets (1ac2) with next (fpu - 0.0).

This solution fits with what we think we'd like from a FPU solution: https://github.com/gcp/leela-zero/pull/827#issuecomment-364214944

gcp commented:
Not looking promising here.

3a2bb61010b15fa06d445f2f387d58461b320133 on better_fpu branch

6665 (strong) and a115 (weak) networks

```
leelaz-next v leelaz-wh (166/1000 games)
board size: 19   komi: 7.5
              wins              black         white        avg cpu
leelaz-next    103 62.05%       41 49.40%     62  74.70%    112.57
leelaz-wh       63 37.95%       21 25.30%     42  50.60%    125.61
                                62 37.35%     104 62.65%

leelaz-next-weak v leelaz-wh-weak (164/1000 games)
board size: 19   komi: 7.5
                   wins              black         white       avg cpu
leelaz-next-weak     70 42.68%       34 41.46%     36 43.90%    415.41
leelaz-wh-weak       94 57.32%       46 56.10%     48 58.54%    418.05
                                     80 48.78%     84 51.22%

```

WhiteHalmos commented:
I think it's safe to say that this change is not strong enough to warrant further consideration. Though it performs okay on average, it regresses our strongest net and flip flops for the weaker nets.

Here are all the test results:

```
6665 (elo 7918) leelaz-next v leelaz-wh (166/1000 games)
board size: 19   komi: 7.5
              wins              black         white        avg cpu
leelaz-next    103 62.05%       41 49.40%     62  74.70%    112.57
leelaz-wh       63 37.95%       21 25.30%     42  50.60%    125.61
                                62 37.35%     104 62.65%

d6f3 (elo 7853): LZ-next v LZ-next-visited-best-eval (121/400 games)
board size: 19   komi: 7.5
                            wins              black         white       avg cpu
LZ-next                       60 49.59%       31 50.82%     29 48.33%    243.71
LZ-next-visited-best-eval     61 50.41%       31 51.67%     30 49.18%    259.75
                                              62 51.24%     59 48.76%

a115 (elo 893): leelaz-next-weak v leelaz-wh-weak (164/1000 games)
board size: 19   komi: 7.5
                   wins              black         white       avg cpu
leelaz-next-weak     70 42.68%       34 41.46%     36 43.90%    415.41
leelaz-wh-weak       94 57.32%       46 56.10%     48 58.54%    418.05
                                     80 48.78%     84 51.22%

1ac2 (elo 543): LZ-next-fpu0 v LZ-next-visited-best-eval (194/400 games)
board size: 19   komi: 7.5
                            wins              black          white       avg cpu
LZ-next-fpu0                 114 58.76%       69  71.13%     45 46.39%    193.66
LZ-next-visited-best-eval     80 41.24%       52  53.61%     28 28.87%    193.62
                                              121 62.37%     73 37.63%
```

Eddh commented:
Too bad :(
